diff --git a/packages/SystemUI/AndroidManifest.xml b/packages/SystemUI/AndroidManifest.xml
index dea0693..f03c100 100644
--- a/packages/SystemUI/AndroidManifest.xml
+++ b/packages/SystemUI/AndroidManifest.xml
@@ -202,6 +202,8 @@
 
     <uses-permission android:name="android.permission.READ_PHONE_STATE" />
 
+    <uses-permission android:name="android.permission.WRITE_APN_SETTINGS" />
+
     <!--SPRD: Bug 692453 customize protected broadcast -->
     <protected-broadcast android:name = "com.action.hide_navigationbar"/>
     <!-- @} -->
diff --git a/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSFactoryImpl.java b/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSFactoryImpl.java
index 13b3ffc..ee8bb19 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSFactoryImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSFactoryImpl.java
@@ -57,6 +57,8 @@ import com.sprd.systemui.SystemUIPluginUtils;
 import com.sprd.systemui.SystemUIAudioProfileUtils;
 import com.android.systemui.statusbar.phone.StatusBar;
 
+import tiles.DataSwitchTile;
+
 
 public class QSFactoryImpl implements QSFactory {
 
@@ -89,6 +91,8 @@ public class QSFactoryImpl implements QSFactory {
         else if (tileSpec.equals("cast")) return new CastTile(mHost);
         else if (tileSpec.equals("hotspot")) return new HotspotTile(mHost);
         else if (tileSpec.equals("user")) return new UserTile(mHost);
+        else if (tileSpec.equals("dataswitch") && isShowDataSwitch())
+            return new DataSwitchTile(mHost);
          //else if (tileSpec.equals("battery")) return new BatterySaverTile(mHost);
         /*SPRD bug 780822:Super power feature*/
         //else if (tileSpec.equals("battery")) return new BatteryTile(this);
@@ -163,4 +167,13 @@ public class QSFactoryImpl implements QSFactory {
         return isVersionShowVolte;
     }
     /* @} */
+
+    private boolean isShowDataSwitch() {
+        boolean isShowDataSwitch = false;
+        if (mHost.getContext() != null) {
+            isShowDataSwitch = mHost.getContext().getResources().getBoolean(R.bool.config_isShowDataSwitch_bool);
+        }
+        Log.d(TAG, "isShowDataSwitch=" + isShowDataSwitch);
+        return isShowDataSwitch;
+    }
 }
