diff --git a/src/java/com/android/internal/telephony/DataEnableController.java b/src/java/com/android/internal/telephony/DataEnableController.java
--- a/src/java/com/android/internal/telephony/DataEnableController.java
+++ b/src/java/com/android/internal/telephony/DataEnableController.java
@@ -154,8 +154,13 @@ public class DataEnableController extends ContextWrapper {
         public DataRegisterContentObserver(Context context, int subId) {
             mContext = context;
             mSubId = subId;
+            String settingsString = Settings.Global.MOBILE_DATA;
+            if (TelephonyManager.getDefault().getSimCount() != 1) {
+                settingsString = Settings.Global.MOBILE_DATA + mSubId;
+            }
+            Log.d(TAG, "listen URI=" + settingsString);
             mContext.getContentResolver().registerContentObserver(
-                    Settings.Global.getUriFor(Settings.Global.MOBILE_DATA + mSubId), true,
+                    Settings.Global.getUriFor(settingsString), true,
                     mMobileDataObserver);
         }

